idx/memory/tag_query.go

```diff
@@ -0,0 +1,411 @@
+		// (f.e. expression "key!=" will be translated to "key=~^.+")
+		if e.value == "^.+" {
+			shortCut = true
+		} else {
+			re, err = regexp.Compile(e.value)
+			if err != nil {
+				return errInvalidQuery
+			}
+		}
+
+		// cache all tags that have once matched the regular expression.
+		// this is based on the assumption that many matching tags will be repeated
+		// over multiple series, so there's no need to run the regex for each of them
+		// because once we know that a tag matches we can just compare strings
+		matchingTags := make(map[string]struct{})
+		notMatchingTags := make(map[string]struct{})
+	IDS:
+		for id := range resultSet {
+			var def *idx.Archive
+			var ok bool
+			if def, ok = byId[id.String()]; !ok {
+				// should never happen because every ID in the tag index
+				// must be present in the byId lookup table
+				CorruptIndex.Inc()
+				log.Error(3, "memory-idx: ID %s is in tag index but not in the byId lookup table", id.String())
+				delete(resultSet, id)
+				continue IDS
+			}
+
+			for _, tag := range def.Tags {
+				// reduce regex matching by looking up cached matches
+				if _, ok := matchingTags[tag]; ok {
+					if not {
+						delete(resultSet, id)
+					}
+					continue IDS
+				}
+			}
+
+			for _, tag := range def.Tags {
+				// length of key doesn't match
+				if len(tag) <= len(e.key)+1 || tag[len(e.key)] != 61 {
+					continue
+				}
+
+				// reduce regex matching by looking up cached non-matches
+				if _, ok := notMatchingTags[tag]; ok {
+					continue
+				}
+
+				if e.key == tag[:len(e.key)] && (shortCut || re.MatchString(tag[len(e.key)+1:])) {
+					if len(matchingTags) < matchCacheSize {
+						matchingTags[tag] = struct{}{}
+					}
+					if not {
+						delete(resultSet, id)
+					}
+					continue IDS
+				} else {
+					if len(notMatchingTags) < matchCacheSize {
+						notMatchingTags[tag] = struct{}{}
+					}
+				}
+			}
+			if !not {
+				delete(resultSet, id)
+			}
+		}
+	}
+	return nil
+}
+
+func (q *TagQuery) filterByFrom(resultSet TagIDs, byId map[string]*idx.Archive) {
+	if q.from <= 0 {
+		return
+	}
+
+	for id := range resultSet {
+		var def *idx.Archive
+		var ok bool
+		if def, ok = byId[id.String()]; !ok {
+			// should never happen because every ID in the tag index
+			// must be present in the byId lookup table
+			CorruptIndex.Inc()
+			log.Error(3, "memory-idx: ID %s is in tag index but not in the byId lookup table", id.String())
+			delete(resultSet, id)
+			continue
+		}
+
+		if def.LastUpdate < q.from {
+			delete(resultSet, id)
+		}
+	}
+}
+
+func (q *TagQuery) Run(index TagIndex, byId map[string]*idx.Archive) (TagIDs, error) {
+	var skipMatch, skipEqual = -1, -1
+	var resultSet TagIDs
+	var err error
+
+	// find the best expression to start with and retrieve its resultSet
+	if len(q.equal) == 0 {
+		if len(q.match) == 0 {
+			return nil, errInvalidQuery
+		}
+
+		skipMatch, resultSet, err = q.getInitialByMatch(index)
+		if err != nil {
+			return nil, errInvalidQuery
```

Comment:
if we can do validation up front, we don't need any error handling here anymore, but if we do, we should probably just return `err` instead of an err that may be something else.

Reply:
:+1: gotten rid of it